### PR TITLE
feat(protocol): update `PlonkVerifier` to accept new public inputs

### LINKED ISSUE #393: feat(protocol): implement ZKP verifier

### PR DESCRIPTION
**Pull request by davidtaikocha**

### CODE REVIEW

**Comment by johntaiko** [comment]
**Attention: It's not the final version, don't approve**

Because the contract of verifier is too large and more than 24kb, so I will optimize the size in next version

**Comment by davidtaikocha** [comment]
> **Attention: It's not the final version, don't approve**
> 
> Because the contract of verifier is too large and more than 24kb, so I will optimize the size in next version

Yah after the verification contract bytecode size issue fixed, i will change this PR's status to "Ready for review".

File: packages/protocol/contracts/libs/LibBlockHeader.sol [hunk 1]
```diff
@@ -34,7 +34,15 @@ library LibBlockHeader {
...
         BlockHeader memory header
     ) internal pure returns (bytes32) {
-        bytes[] memory list;
+        bytes memory rlpHeader = LibRLPWriter.writeList(
+            getBlockHeaderRLPItemsList(header)
+        );
+        return keccak256(rlpHeader);
+    }
+
+    function getBlockHeaderRLPItemsList(
```
Comment (dantaik) [inline packages/protocol/contracts/libs/LibBlockHeader.sol:43]: make it private and rename to `_getBlockHeaderRLPItemsList`?

**Review by dantaik — DISMISSED** [review]

File: packages/protocol/contracts/L1/libs/LibProving.sol [hunk 2]
```diff
@@ -459,4 +451,24 @@ library LibProving {
             header.mixHash != meta.mixHash
         ) revert L1_META_MISMATCH();
     }
+
+    function _getInstance(
+        Evidence memory evidence
+    ) internal pure returns (bytes32 instance) {
```
Comment (dantaik) [inline packages/protocol/contracts/L1/libs/LibProving.sol:457]: @smtmfft @davidtaikocha I'm wonder, if the circuits really need all inputs to be RLP encoded first, won't packing different bytes32 together be more efficient?

Comment (johntaiko) [reply packages/protocol/contracts/L1/libs/LibProving.sol:457]: I think there are two reasons:
1. RLP encoding can be used for more complex data structures, such as list, nested list, etc. (Assuming pi will change in the future
2. The main reason ABI encoding and RLP encoding will require two different circuits, but they actually do the same thing. RLP is more versatile than ABI with a little overhead. More circuits make optimization difficult.

File: packages/protocol/contracts/libs/LibBlockHeader.sol (hunk 1 above)
Comment (davidtaikocha) [reply packages/protocol/contracts/libs/LibBlockHeader.sol:43]: we still need it to be `internal` since Its also used by `LibProving._getInstance`?

**Review by dantaik — APPROVED** [review]

**Review by johntaiko — APPROVED** [review]

**Pull request merged**